### PR TITLE
Implemented QC variable name matching data variable name test, as wel…

### PR DESCRIPTION
**Pull request by ggalibert**
…l as related unit test.

### CODE REVIEW

File: compliance_checker/imos/imos.py [hunk 1]
```diff
@@ -1311,6 +1311,29 @@ def check_quality_control_conventions_for_quality_control_variable(self, dataset
...
         return ret_val
 
+    def check_quality_control_variable_matches_variable(self, dataset):
+        """
+        Check that the name of a quality control variable which is <DATA>_quality_control
+        matches the name of a corresponding data variable <DATA>.
+        """
+        ret_val = []
+        for qc_variable in self._quality_control_variables:
+            qc_variable_root_name = qc_variable.name[:-16] # '_quality_control' is 16 characters long
```
Comment (mhidas) [inline compliance_checker/imos/imos.py]: The first thing we want to check is that the variable name actually ends in '_quality_control'. e.g.

``` python
result_name = ('qc_var', qc_variable.name, 'name')
reasoning = []
match = re.findall('^(.*)_quality_control$', name)   # returns a list with the root name, if matched
if not match:
    reasoning = ["Name of QC variable should end in '_quality_control'"]
ret_val.append( Result(BaseCheck.HIGH, match==[], result_name, reasoning) )
```

Comment (mhidas) [inline compliance_checker/imos/imos.py]: ... then we can `continue` if it didn't match, or `qc_variable_root_name = match[0]`

File: compliance_checker/imos/imos.py [hunk 2]
```diff
@@ -1311,6 +1311,29 @@ def check_quality_control_conventions_for_quality_control_variable(self, dataset
...
+            result_name = ('qc_var', qc_variable.name, 'check_matching_variable_name')
```
Comment (mhidas) [inline compliance_checker/imos/imos.py]: I've been removing the redundant word `check` from the check names, trying to keep them relatively short. How about `('qc_var', qc_variable.name, 'match_variable_name')`?

File: compliance_checker/imos/imos.py [hunk 3]
```diff
@@ -1311,6 +1311,29 @@ def check_quality_control_conventions_for_quality_control_variable(self, dataset
...
+            reasoning = ["there is no data variable name '%s' for '%s'" % (qc_variable_root_name, qc_variable.name)]
+            match = False
+            for name, var in dataset.dataset.variables.iteritems():
+              if qc_variable_root_name == name:
```
Comment (mhidas) [inline compliance_checker/imos/imos.py]: Please use 4 spaces for every indent.

File: compliance_checker/imos/imos.py [hunk 4]
```diff
@@ -1311,6 +1311,29 @@ def check_quality_control_conventions_for_quality_control_variable(self, dataset
...
+        Check that the name of a quality control variable which is <DATA>_quality_control
+        matches the name of a corresponding data variable <DATA>.
+        """
+        ret_val = []
+        for qc_variable in self._quality_control_variables:
+            qc_variable_root_name = qc_variable.name[:-16] # '_quality_control' is 16 characters long
+            result_name = ('qc_var', qc_variable.name, 'check_matching_variable_name')
+            reasoning = ["there is no data variable name '%s' for '%s'" % (qc_variable_root_name, qc_variable.name)]
+            match = False
+            for name, var in dataset.dataset.variables.iteritems():
```
Comment (mhidas) [inline compliance_checker/imos/imos.py]: You don't really need a loop here. You can just say 

``` python
if qc_variable_root_name in dataset.dataset.variables.keys():
```

File: compliance_checker/imos/imos.py (hunk 1 above)
Comment (ggalibert) [inline compliance_checker/imos/imos.py]: Isn't checking this redundant? 

`self._quality_control_variables` already returns the list of variables ending with '_quality_control'. Maybe I could test that this list is not empty before continuing.

File: compliance_checker/tests/test_imos.py [hunk 5]
```diff
@@ -680,6 +680,14 @@ def test_check_data_variables(self):
         for i in range(1,5):
             self.assertFalse(ret_val[i].value)
 
+    def test_check_quality_control_variable_matches_variable(self):
+        self.imos.setup(self.test_variable_dataset)
+        ret_val = self.imos.check_quality_control_variable_matches_variable(self.test_variable_dataset)
+        self.assertTrue(ret_val != None)
+        self.assertTrue(len(ret_val) == 4)
```
Comment (mhidas) [inline compliance_checker/tests/test_imos.py]: If one of these fails, you'll get more helpful output if you use more specific assert statements (see [unittest documentation](https://docs.python.org/2/library/unittest.html#test-cases):

``` python
    self.assertIsNotNone(ret_val)
    self.assertEqual(len(ret_val), 4)
```

**Comment by mhidas** [comment]
@ggalibert Looks good, but I think the first part of the check is still missing. See my inline comments.

File: compliance_checker/imos/imos.py (hunk 4 above)
Comment (ggalibert) [inline compliance_checker/imos/imos.py]: Ok, this works. I'll include it.

File: compliance_checker/tests/test_imos.py (hunk 5 above)
Comment (ggalibert) [inline compliance_checker/tests/test_imos.py]: Sounds good. I'll do that, when this current PR is merged, in a separate commit and PR which will apply to the whole test_imos.py.

**Comment by ggalibert** [comment]
@mhidas, ready for further review/discussion. 

I've included most of your comments, except the '_quality_control' check at first which I thought was redundant. Indeed, self._quality_control_variables already returns the list of variables ending with '_quality_control' only, if any.

File: compliance_checker/tests/test_imos.py (hunk 5 above)
Comment (mhidas) [inline compliance_checker/tests/test_imos.py]: :+1:

**Comment by ggalibert** [comment]
@mhidas , I took your comments into account. Should be ready for another review. Had to update unittest and test netcdf file accordingly so that the two checks are tested.

Added a qc var in imos_variable_test.nc such as :

```
byte bad2_qc ;
    bad2_qc:standard_name = "data_variable status_flag" ;
```

**Comment by mhidas** [comment]
Great! Thanks @ggalibert

**Pull request merged**